### ISSUE TITLE
Signed cookies not available in controller tests

### ISSUE THREAD

**Issue by sgrif** [rails/rails#27145]
This issue has been ported from the report at https://github.com/rspec/rspec-rails/issues/1658, as it can be replicated purely with Rails using public API. This test passes using Rails 4.2

### Steps to reproduce

<details>
<summary>Reproduction Script</summary>

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = File.dirname(__FILE__)
  secrets.secret_token    = "secret_token"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    render plain: "Home"
  end

  def foo_cookie
    cookies.encrypted["foo"]
  end
end

require "rails/test_help"
require "minitest/autorun"

class TestControllerTest < ActionController::TestCase
  setup do
    cookies.encrypted["foo"] = "bar"
  end

  test "accessing cookies directly" do
    get :index
    assert_equal "bar", cookies.encrypted["foo"]
  end

  test "accessing helper method" do
    get :index
    assert_equal "bar", @controller.foo_cookie
  end
end
```

</details>

### Expected behavior
Changes to `cookies.encrypted` in the test setup should appear in the controller's cookies

### Actual behavior
Changes to `cookies.encrypted` in the test setup do not affect the controller's cookies

### System configuration
**Rails version**: 5.0.0, 5.0.0.1, master

**Ruby version**: 2.3.3

**Comment by maclover7** [comment]
Looks like this was introduced via https://github.com/rails/rails/commit/ae29142142324545a328948e059e8b8118fd7a33, cc @eileencodes @tenderlove

**Comment by maclover7** [comment]
I *think* I have a fix for this, just adding a regression test to the test suite.

**Comment by maclover7** [comment]
Opened #27146 with a possible solution to this ticket.

**Comment by maclover7** [comment]
Closing since #27146 was merged.

**Comment by devohh** [comment]
@maclover7 @sgrif I think the problem is still alive. I left a comment on [that page](https://github.com/rails/rails/pull/27586#issuecomment-282965863). Am I missing something ?

**Comment by rafaelfranca** [comment]
@devohh this is still an issue that needs to be fixed.

**Comment by eileencodes** [comment]
Hrm, the original issue was about controller tests, and @devohh said the problem is with integration tests? Can we get a test script demonstrating the problem?

**Comment by eileencodes** [comment]
Oh sorry @rafaelfranca I didn't see you reverted the original fix. Nevermind :)

**Comment by devohh** [comment]
@eileencodes @rafaelfranca The same issue occurs during integration test AND also during controller test.

**Comment by alexcameron89** [comment]
This issue has been fixed in Rails 5.1 & master by 78c6c4b24810e505921d0bb3575c7a59e416c268 from @maclover7.

To double check, I also verified it with the original reproduction app in the RSpec issue, and it passed with an upgrade to 5.1.0. :+1:

**Comment by eileencodes** [comment]
That commit from Jon was merged but there was a second that was related and reverted, but never fixed. See #27586

**Comment by alexcameron89** [comment]
Sorry about that, I did not run the script from #27586. Thanks @eileencodes. 

Here's the second script mentioned for reference on this issue.

``` ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails", branch: "5-0-stable"
  gem "pry-byebug"
end

require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = File.dirname(__FILE__)
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_token    = "secret_token"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    resources :users, only: [:index]
  end
end

class UsersController < ActionController::Base
  def index
    unless cookies[:name].present?
      cookies[:name] = "Alice"
    end
    render plain: "Home"
  end
end

require "minitest/autorun"
require "rack/test"

class UsersControllerTest < ActionController::TestCase
  setup do
    @controller = UsersController.new
    @routes = Rails.application.routes
  end

  test '#index sets a cookie if it is not already present' do
    get :index
    assert_response :ok
    assert_equal "Alice", cookies[:name]

    cookies[:name] = "Bob"

    get :index
    assert_response :ok
    assert_equal "Bob", cookies[:name]
  end
end

**Comment by schneems** [comment]
This issue is still valid?

**Comment by denisw** [comment]
On a fresh Rails 5.2 install, I still can't use `cookies.encrypted` in an integration test:

```ruby
require 'test_helper'

class MyControllerTest < ActionDispatch::IntegrationTest
  test "displays current user's first name" do
    user = users(:alice)
    cookies.encrypted[:current_user] = user.id
    get '/'
    assert_select "p", user.first_name
  end
end
```

Results in:

```
Error:
MyControllerTest#test_displays_current_user's_first_name:
NoMethodError: undefined method `encrypted' for #<Rack::Test::CookieJar:0x00007fea2f04ad00>
    test/controllers/my_controller_test.rb:6:in `block in <class:DashboardControllerTest>'
```

**Comment by vtm9** [comment]
@eileencodes I have a similar bug. `cookies.encrypted[]` don't decrypt value.
In dev and prod envs if you put something in `cookies.encrypted[:user_session] = 'session_token'`, then
 `cookies.encrypted[:user_session]` will returns decrypted value `'session_token'` .
But in the test env in `ActionDispatch::IntegrationTest`, if you call `cookies.encrypted[:user_session]` it will returns encrypted value like `cookies[:user_session]` `"VFJPRlN(...)UTzZ--5d55a9ad0a51de5045330df2954a0b90e6f9afaf` 
Rails 5.2.1

**Comment by shicholas** [comment]
Can confirm too, I guess an alt is to log in or fetch a token for an API first instead of trying to stub the cookie before the test.

**Comment by rafaelfranca** [comment]
This issue is fixed for `ActionController::TestCase` based tests. `ActionDispatch::IntegrationTest` doesn't support access to encrypted cookies by design because they are integration tests and don't allow access to anything that a browser would not.

**Comment by oreoshake** [comment]
I thought `ActionController::TestCase` is (was?) deprecated?

**Comment by gobijan** [comment]
With Rails 8 Beta 1 this might get relevant again.

In the new authentication scope we have:

 def find_session_by_cookie
    Session.find_by(id: cookies.signed[:session_id])
  end
    
 But ActionDispatch::Cookies::CookieJar doesn't define signed.
 
 I monkey patched in my test helpers like it's 2009:
 
 ENV["RAILS_ENV"] ||= "test"
require_relative "../config/environment"
require "rails/test_help"

module ActiveSupport
  class TestCase
    # Run tests in parallel with specified workers
    parallelize(workers: :number_of_processors)

    # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
    fixtures :all

    # Add more helper methods to be used by all tests here...
    def login_as(user)
      session = user.sessions.create!(user_agent: "test", ip_address: "127.0.0.1")
      Current.session = session
      cookies[:session_id] = session.id
    end

    def encrypted_password(password = "password")
      BCrypt::Password.create(password, cost: 4)
    end

    def assert_logged_in
      assert Current.session.present?, "Expected to be logged in, but wasn't"
    end

    def assert_not_logged_in
      assert_not Current.session.present?, "Expected not to be logged in, but was"
    end
  end
end

# Monkey-patch the Authentication concern for test environment
module Authentication
  private
    def find_session_by_cookie
      Session.find_by(id: cookies[:session_id])
    end
end


Maybe we do something here? What would be the desired outcome?

**Comment by rafaelfranca** [comment]
Can you provide an example of what doesn't work so we can run and open a new issue? `ActionDispatch::Cookies::CookieJar` for sure defines `signed`

```ruby
>> cookies = ActionDispatch::Cookies::CookieJar.build({}, {})
=> #<ActionDispatch::Cookies::CookieJar:0x0000ffff6b4ac868 @committed=false, @cookies={}, @delete_cookies={}, @request={}, @set_cookies={}>
>> cookies.method(:signed)
=> #<Method: ActionDispatch::Cookies::CookieJar(ActionDispatch::Cookies::ChainedCookieJars)#signed() /workspaces/rails/actionpack/lib/action_dispatch/middleware/cookies.rb:253>
```

**Comment by gobijan** [comment]
Sorry Rafael, I was overworked and out of sleep. I meant Rack::Test::CookieJar has no signed instance method (corrected my comment).
With the monkey patch I worked around that. I'll create an issue.

**Comment by aurelienbottazini** [comment]
I monkey patched a little differently with Rails 8

my test_helper.rb

I chose to have a session by default
I have an helper to test some controller without a session
Did not want to involve the db at all to keep it as fast as I can

```ruby
ENV["RAILS_ENV"] ||= "test"
require_relative "../config/environment"
require "rails/test_help"

module ActiveSupport
  class TestCase
    # Run tests in parallel with specified workers
    parallelize(workers: :number_of_processors)

    # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
    fixtures :all

    setup do
      $test_session = Session.new(user: users(:one))
    end

    def without_session
      $test_session = nil
      yield
      $test_session = Session.new(user: users(:one))
    end
  end
end

module Authentication
 private
   def find_session_by_cookie
     $test_session
   end
end
```

Helper usage
```ruby
require "test_helper"

class PublicControllerTest < ActionDispatch::IntegrationTest
  test "should get index" do
    without_session do
      get public_index_path
      assert_redirected_to new_session_path
    end
  end
end
```
